src/github-client-finder/file-finder.ts

```diff
@@ -0,0 +1,62 @@
+import * as ghCore from "@actions/core";
+import * as semver from "semver";
+
+import { ClientFile, InstallableClient } from "../util/types";
+import { getOS, getArch, getGitHubReleaseAssetPath } from "../util/utils";
+import { findClientVersionFromGithub, getReleaseAssets } from "./repository-finder";
+import {
+    filterByOS, filterByZipped, filterByArch, filterClients, filterByExecutable,
+} from "../util/filters";
+
+type ClientFilterFunc = ((filename: string) => boolean);
+
+/**
+ * Finds the latest version of client that satisifies the desiredVersionRange.
+ * This is done by:
+ *  - looking up the github releases for the client on the respective github repositories
+ *  - reading the available versions (releases)
+ *  - finding the max version that satisfies
+ *  - navigating into that version/release
+ *  - finding the filename/asset for the client, operating system, and architecture for the current machine.
+ *
+ * @returns All the required information about the client file, once it's been located.
+ */
+export async function findMatchingClientFromGithub(client: InstallableClient, desiredVersionRange: semver.Range):
```

Comment:
this is _very_ similar to the mirror version of this function. can you try and combine the two, with a toggle for github vs mirror?